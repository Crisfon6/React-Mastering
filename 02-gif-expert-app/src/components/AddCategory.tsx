import { useState } from "react";

/* export function AddCategory({setCategories, categories}: {setCategories: (value: string[] | ((prevState: string[]) => string[])) => void, categories: string[]}) {
    const [input,setInput] = useState('');
    const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault();
        const form = e.target as HTMLFormElement;
        const inputValue:string = form.search.value;
        if(categories.includes(inputValue.trim()) || inputValue.trim() === '') return;
        setCategories((cat:string[])=>([inputValue.trim(),...cat]));
        form.reset();
    }
    return (
        <form onSubmit={handleSubmit}>
            <input type="text" placeholder="Search gifs" id="search" value={input} onChange={(e)=>setInput(e.target.value)}/>
            <button type="submit">Add</button>            
        </form>
    )
} */
export function AddCategory({onNewCategory}: {onNewCategory: (newCategory:string)=>void}) {
    const [input,setInput] = useState('');
    const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => { 
        e.preventDefault();
        const cleanInput = input.trim();
        if(cleanInput.length <= 1) return;
        onNewCategory(cleanInput);
        setInput('');
    }
    return (
        <form onSubmit={handleSubmit} className="flex flex-col sm:flex-row gap-2 mb-6 w-full max-w-2xl" aria-label="form">
            <input 
                type="text" 
                placeholder="Search gifs" 
                id="search" 
                value={input} 
                onChange={(e)=>setInput(e.target.value)}
                className="flex-grow px-4 py-2 rounded-lg border border-gray-300 focus:border-purple-500 focus:ring-2 focus:ring-purple-200 outline-none transition-all"
            />
            <button 
                type="submit"
                className="block sm:hidden bg-purple-300 hover:bg-purple-400 text-white font-medium px-6 py-2 rounded-lg transition-colors shadow-md hover:shadow-lg cursor-pointer"
            >
                Add
            </button>            
        </form>
    )
}
