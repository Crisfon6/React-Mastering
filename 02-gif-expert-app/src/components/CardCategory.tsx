export function CardCategory({url,title}:{url:string,title:string}){
    return(
        <div className="hover:animate-background rounded-xl bg-gradient-to-r from-green-300 via-blue-500 to-purple-600 p-0.5 shadow-xl transition hover:bg-[length:400%_400%] hover:shadow-xs hover:[animation-duration:_4s] hover:scale-105 hover:shadow-lg">
        <div className="rounded-[10px] bg-white p-4 ">
        <div data-testid="gif-image" style={{ backgroundImage: `url(${url})` }} className="bg-cover bg-center h-48" ></div>
          <h1 className="text-2xl font-bold mb-6 text-center text-rose-50 tracking-wide drop-shadow-[0_2px_2px_rgba(0,0,0,0.8)] py-1 px-1    transform hover:scale-105 transition-transform duration-300 ">
            {title}
          </h1>
        </div>
      </div>
    )
}
