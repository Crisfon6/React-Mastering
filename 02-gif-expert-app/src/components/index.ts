export * from "./AddCategory";
export * from "./CardCategory";
export * from "./GridCategory";
