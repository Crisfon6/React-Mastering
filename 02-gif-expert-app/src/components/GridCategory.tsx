import { CardCategory } from "./CardCategory";
import { GiftSimple } from "../models/Gift";
import { useFetchGifs } from "../hooks/useFetchGifs";

export function GridCategory({ category }: { category: string }) {
  const { gifs, isLoading } = useFetchGifs(category);

  return (
    <>
      <div className="px-10">
        {isLoading && (
          <div className="flex items-center justify-center min-h-[200px]">
            <div className="animate-spin rounded-full h-16 w-16 border-t-4 border-purple-500 border-solid"></div>
            <p className="ml-4 text-xl text-rose-50 font-semibold"> </p>
          </div>
        )}
        <h1 className="pt-2 text-3xl md:text-5xl lg:text-5xl font-bold  text-left text-rose-50 tracking-wide [text-shadow:_-1px_-1px_0_#a855f7,_1px_-1px_0_#a855f7,_-1px_1px_0_#a855f7,_1px_1px_0_#a855f7] py-1 px-1 transform ">
          {category.toUpperCase()}
        </h1>

        <div className="grid grid-cols-1 sm:grid-cols-2 md:grid-cols-3 lg:grid-cols-4 gap-4 pt-4 cursor-pointer">
          {gifs.map((gif: GiftSimple) => (
            <CardCategory key={gif.id} gif={gif}></CardCategory>
          ))}
        </div>
      </div>
    </>
  );
}
