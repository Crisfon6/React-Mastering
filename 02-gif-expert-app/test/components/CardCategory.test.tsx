import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { CardCategory } from "../../src/components/CardCategory";
import React from "react";

describe("CardCategory", () => {
  test("to match snapshot", () => {
    const url = "https://media1.giphy.com/media/v1.Y2lkPTNhZGQxM2Vidmw2eWtvYnJjNGJuN3g1N2F6anBlbThjc3ByNWZtOGdsejhlN2s5MSZlcD12MV9naWZzX3NlYXJjaCZjdD1n/VXJWhaO7afRe/giphy.gif";

    const {container} =  render(<CardCategory title="test" url={url}/>);
    expect(container).toMatchSnapshot();
  });
  test("should render the card category", () => {
    const url = "https://media1.giphy.com/media/v1.Y2lkPTNhZGQxM2Vidmw2eWtvYnJjNGJuN3g1N2F6anBlbThjc3ByNWZtOGdsejhlN2s5MSZlcD12MV9naWZzX3NlYXJjaCZjdD1n/VXJWhaO7afRe/giphy.gif";

    render(<CardCategory title="test" url={url}/>);
    expect(screen.getByText("test")).toBeTruthy();
  });
  test("should render the card category with the correct url", () => {
    const url = "https://media1.giphy.com/media/v1.Y2lkPTNhZGQxM2Vidmw2eWtvYnJjNGJuN3g1N2F6anBlbThjc3ByNWZtOGdsejhlN2s5MSZlcD12MV9naWZzX3NlYXJjaCZjdD1n/VXJWhaO7afRe/giphy.gif";

    render(<CardCategory title="test" url={url}/>);
    const backgroundDiv = screen.getByTestId("gif-image");
    expect(backgroundDiv).toHaveStyle(`background-image: url(${url})`);
  });
  test("should render the card category with the correct title", () => {
    const title = "test";
    const url = "https://media1.giphy.com/media/v1.Y2lkPTNhZGQxM2Vidmw2eWtvYnJjNGJuN3g1N2F6anBlbThjc3ByNWZtOGdsejhlN2s5MSZlcD12MV9naWZzX3NlYXJjaCZjdD1n/VXJWhaO7afRe/giphy.gif";
    render(<CardCategory title={title} url={url}/>);
    expect(screen.getByText(title)).toBeTruthy();
    expect(screen.getByRole("heading", {name: title}).innerHTML).toBe(title);
  });
});
