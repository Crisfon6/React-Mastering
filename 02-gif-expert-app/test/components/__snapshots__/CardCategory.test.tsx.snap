// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardCategory to match snapshot 1`] = `
<div>
  <div
    class="hover:animate-background rounded-xl bg-gradient-to-r from-green-300 via-blue-500 to-purple-600 p-0.5 shadow-xl transition hover:bg-[length:400%_400%] hover:shadow-xs hover:[animation-duration:_4s] hover:scale-105 hover:shadow-lg"
  >
    <div
      class="rounded-[10px] bg-white p-4 "
    >
      <div
        class="bg-cover bg-center h-48"
        data-testid="gif-image"
        style="background-image: url(https://media1.giphy.com/media/v1.Y2lkPTNhZGQxM2Vidmw2eWtvYnJjNGJuN3g1N2F6anBlbThjc3ByNWZtOGdsejhlN2s5MSZlcD12MV9naWZzX3NlYXJjaCZjdD1n/VXJWhaO7afRe/giphy.gif);"
      />
      <h1
        class="text-2xl font-bold mb-6 text-center text-rose-50 tracking-wide drop-shadow-[0_2px_2px_rgba(0,0,0,0.8)] py-1 px-1    transform hover:scale-105 transition-transform duration-300 "
      >
        test
      </h1>
    </div>
  </div>
</div>
`;
